### PR TITLE
Enh re exception handling

### PR DESCRIPTION
**Pull request by tacaswell**
This still needs 

  - more evil tests
  - There is now a path to `halt` and for a plan to kill the RE dead as `GeneratorExit` is not a sub-class of `Exception` for exactly this reason.

### CODE REVIEW

**Comment by tacaswell** [comment]
@danielballan @klauer I think this is ready to review now.  I suggest doing it commit by commit, I left nice messages.

Closes #502 

This also should fix the issue where stop hangs on long exposures an CSX (@ambarb I know you have reported this verbally, but I can not find an issue)

File: bluesky/run_engine.py [hunk 1]
```diff
@@ -906,7 +906,8 @@ def _run(self):
             self._exit_status = 'success'
             # TODO Is the sleep here necessasry?
             yield from asyncio.sleep(0.001, loop=self.loop)
-        except (FailedPause, RequestAbort, asyncio.CancelledError):
+        except (FailedPause, RequestAbort, asyncio.CancelledError,
```
Comment (danielballan) [inline bluesky/run_engine.py]: I think 'fail' is a better fit than 'abort'.

Comment (tacaswell) [inline bluesky/run_engine.py]: This can be injected through `halt` which is more of an abort?

File: bluesky/run_engine.py [hunk 2]
```diff
@@ -894,16 +905,20 @@ def _run(self):
                         # exceptions (coming in via throw) can be
                         # raised
                         response = yield from coro(msg)
+                    # spacial case `CancelledError` and let the outer
```
Comment (danielballan) [inline bluesky/run_engine.py]: spelling

File: bluesky/run_engine.py [hunk 3]
```diff
@@ -808,6 +808,7 @@ def _run(self):
         - Try to remove any monitoring subscriptions left on by the plan.
         - If interrupting the middle of a run, try to emit a RunStop document.
         """
+        is_canceled = None
```
Comment (danielballan) [inline bluesky/run_engine.py]: spelling

Comment (tacaswell) [inline bluesky/run_engine.py]: rename to canceled exception

Comment (tacaswell) [inline bluesky/run_engine.py]: pending_cancel_exception

File: bluesky/run_engine.py [hunk 4]
```diff
@@ -786,6 +786,18 @@ def stop(self):
...
             self._resume_event_loop()
 
+    def halt(self):
+        '''Stop the running plan and do not allow the plan a chance to clean up
+        '''
+        if self.state.is_idle:
+            raise TransitionError("RunEngine is already idle.")
+        print("HALTING...")
+        self._interrupted = True
+        self._exception = GeneratorExit()
```
Comment (tacaswell) [inline bluesky/run_engine.py]: wrap in custom layer

**Comment by danielballan** [comment]
Update all three print messages to help users remember the differences:

Aborting: running cleanup and marking exit_status as 'abort'...
Stopping: running cleanup and marking exit_status as 'success'...
Halting: skipping cleanup and marking exit_status as 'abort'...

**Comment by tacaswell** [comment]
Made all changes, I think this is good to go :smile:

File: bluesky/run_engine.py [hunk 5]
```diff
@@ -808,75 +828,117 @@ def _run(self):
...
+                                self._exception)
+                        except Exception as e:
+                            # The current plan did not handle it,
+                            # maybe the next plan (if any) would like
+                            # to try
+                            self._plan_stack.pop()
+                            if len(self._plan_stack):
+                                self._exception = e
+                                continue
+                            # no plans left and still an UN-handled
```
Comment (danielballan) [inline bluesky/run_engine.py]: And still an UN-handled <exception>

File: bluesky/run_engine.py [hunk 6]
```diff
@@ -808,75 +828,117 @@ def _run(self):
...
+                        continue
+
+                    # try to finally run the command the user asked for
+                    try:
+                        # this is one of two places that 'async'
+                        # exceptions (coming in via throw) can be
+                        # raised
                         response = yield from coro(msg)
-                        self._response_stack.append(response)
+                    # spacial case `CancelledError` and let the outer
```
Comment (danielballan) [inline bluesky/run_engine.py]: Bump spelling

File: bluesky/run_engine.py (hunk 5 above)
Comment (tacaswell) [inline bluesky/run_engine.py]: It is a very diplomatic exception

**Comment by danielballan** [comment]
:party:

**Comment by danielballan** [comment]
It's worth saying, thanks for taking a very, very careful approach and sorting this out. This was really great and really important work.

**Pull request merged**